In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

In [23]:
df = pd.read_csv("online_retail_II.csv")

In [24]:
df.shape

(1067371, 8)

In [25]:
df.dtypes

Invoice         object
StockCode       object
Description     object
Quantity         int64
InvoiceDate     object
Price          float64
Customer ID    float64
Country         object
dtype: object

In [26]:
df.isnull().sum()

Invoice             0
StockCode           0
Description      4382
Quantity            0
InvoiceDate         0
Price               0
Customer ID    243007
Country             0
dtype: int64

In [27]:
# remove cancelled orders
df = df[~df['Invoice'].str.startswith('C')]



In [28]:
print("Cancelled orders:", df['Invoice'].str.startswith('C').sum())
print("Total rows:", len(df))

Cancelled orders: 0
Total rows: 1047877


Cancelled transactions(identified by Invoice numbers starting with 'C') were removed, reducing that dataset from 1,067,371 to 1,047,877 rows, ensuring accurate sales and customer analysis.

In [29]:
# drop null values in Customer ID
df = df.dropna(subset=['Customer ID'])

print("Remaining nulls:", df['Customer ID'].isnull().sum())
print("Total rows:", len(df))

Remaining nulls: 0
Total rows: 805620


Customer ID is essential for analysing customer behavior. Rows without Customer ID cannot be linked to any customer, so they were removed to ensure accurate customer-level insights.

In [30]:
# remove rows where Quantity and Price <= 0
df = df[(df['Quantity'] > 0) & (df['Price'] > 0)]

print("Invalid Quantity rows:", (df['Quantity'] <= 0).sum())
print("Invalid Price rows:", (df['Price'] <= 0).sum())
print("Total rows:", len(df))

Invalid Quantity rows: 0
Invalid Price rows: 0
Total rows: 805549


Removed rows where Quantity and Price <= 0, to eliminate returns and erroneous transactions, ensuring accyrate revenue analysis.

In [31]:
print(df['Quantity'].min())
print(df['Price'].min())

1
0.001


In [32]:
print("Total rows:", len(df))

Total rows: 805549


In [33]:
# Fix datatypes
df['InvoiceDate'] = pd.to_datetime(df['InvoiceDate'])

df['Customer ID'] = df['Customer ID'].astype(int).astype(str)

In [35]:
df.dtypes

Invoice                object
StockCode              object
Description            object
Quantity                int64
InvoiceDate    datetime64[ns]
Price                 float64
Customer ID            object
Country                object
dtype: object

In [37]:
# Create revenue column
df['Revenue'] = df['Quantity'] * df['Price']

In [39]:
df[['Quantity', 'Price', 'Revenue']].head()

,Quantity,Price,Revenue
0,12,6.95,83.4
1,12,6.75,81.0
2,12,6.75,81.0
3,48,2.10,100.8
4,24,1.25,30.0


In [45]:
# Extract time features
df['Year'] = df['InvoiceDate'].dt.year
df['Month'] = df['InvoiceDate'].dt.month
df['Month_Name'] = df['InvoiceDate'].dt.strftime('%b')
df['Week'] = df['InvoiceDate'].dt.isocalendar().week
df['DayOfWeek'] = df['InvoiceDate'].dt.day_name()

In [46]:
df[['InvoiceDate', 'Year', 'Month', 'Month_Name', 'Week', 'DayOfWeek']].head()

,InvoiceDate,Year,Month,Month_Name,Week,DayOfWeek
0,2009-12-01 07:45:00,2009,12,Dec,49,Tuesday
1,2009-12-01 07:45:00,2009,12,Dec,49,Tuesday
2,2009-12-01 07:45:00,2009,12,Dec,49,Tuesday
3,2009-12-01 07:45:00,2009,12,Dec,49,Tuesday
4,2009-12-01 07:45:00,2009,12,Dec,49,Tuesday


In [48]:
df.to_csv('retail_clean.csv', index=False)
print("File Saved")

File Saved
